cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCorePeerIdentityHandler.java

```diff
@@ -0,0 +1,74 @@
+/*******************************************************************************
+ * Copyright (c) 2019 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Rikard Höglund (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import org.eclipse.californium.core.coap.Message;
+import org.eclipse.californium.elements.AddressEndpointContext;
+import org.eclipse.californium.elements.EndpointContext;
+import org.eclipse.californium.elements.auth.OSCorePeerIdentity;
+
+/**
+ * Class that handles the EndpointContext for OSCORE messages. 
+ *
+ * It provides functionality to set an appropriate OSCORE peer identity
+ * for an EndpointContext.
+ *
+ */
+public class OSCorePeerIdentityHandler {
+
+	/**
+	 * Updates the currently set endpoint contexts adding an OSCORE peer identity.
+	 * Will only update endpoint contexts that are set as type AddressEndpointContext
+	 * and have an empty or OSCORE peer identity.
+	 * 
+	 * If for instance Scandium has set another type of EndpointContext or another
+	 * type of Principal as peer identity it will not be updated.
+	 * 
+	 * @param oscoreCtx the OSCORE context used for this message
+	 * @param message the message to set the endpoint context for
+	 */
+	public static void setPeerIdentity(OSCoreCtx oscoreCtx, Message message) {
```

Comment:
Are you sure, that changing the destination and source context of the same message makes sense? FMPOV, that is some stuff related to security. So I would prefer to have sendMessage and receiveMessage (or may be sendRequest, sendResponse, receiveRequest, and receiveResponse)

Reply:
My understanding was that by the time a request or response reaches the ObjectSecurityLayer and the OSCORE code it will already have gotten a source or destination endpoint context set. So therefore I am changing it.

But in practice I create the new one with the exact same values as the old one except with an OSCorePeerIdentity as Principal. Basically if an endpoint context is not null I re-create it with the same values but setting a peer identity for OSCORE. 

For every message I will only change either the source or destination endpoint context (depending on which is set) because one of them will be null. At least that is how I understand things from testing the current code.

Reply:
If a message is sent, the destination context is used, the source will be null.
Now you fill it.
If a message is received, the source context is used, the destination will be null.
Now you fill it.
So, please don't fill the context just because it's null. use the destination for sending and source for receiving. That is distinguished to make it clear, what is used/checked and what not.

Reply:
> If a message is sent, the destination context is used, the source will be null.
> Now you fill it.
> If a message is received, the source context is used, the destination will be null.
> Now you fill it.
> So, please don't fill the context just because it's null. use the destination for sending and source for receiving. That is distinguished to make it clear, what is used/checked and what not.

Let me elaborate. I do not fill or touch the context if it is null, due to:
`if(dstContext instanceof AddressEndpointContext &&`
(Which will fail if dstContext is null)
So I actually only update it when it is not null. But I want the peer identity principal in the context to either be null (unset) or OSCorePeerIdentity (set by OSCORE) for me to update it. (To avoid changing things set by Scandium.)

I can certainly change the structure of things so they are more clear what is happening during sending or receiving.

Reply:
OK. I was mislead/mixed up by the null check for the principal.